File: packages/maencof/skills/maencof-craft-skill/scripts/deployment_helper.py
#!/usr/bin/env python3
"""
Intelligent Deployment Helper - Assists with skill deployment decisions

Usage:
    deployment_helper.py <skill-path> [--scope project|global]
    deployment_helper.py --analyze <package.zip>

Features:
- Intelligent deployment location detection
- Project-specific vs global skill recommendations
- Symlink creation support
- Dependency checking
"""

import argparse
import json
import shutil
import subprocess
import sys
import zipfile
from pathlib import Path
from typing import Dict, Optional


def get_claude_skills_dir() -> Path:
    """Get the global Claude skills directory"""
    home = Path.home()
    return home / ".claude" / "skills"


def get_project_skills_dir() -> Optional[Path]:
    """Get the project-specific skills directory if in a project"""
    cwd = Path.cwd()

    # Look for .claude directory in current or parent directories
    current = cwd
    for _ in range(5):  # Check up to 5 levels up
        claude_dir = current / ".claude"
        if claude_dir.exists():
            skills_dir = claude_dir / "skills"
            return skills_dir

        if current.parent == current:  # Reached root
            break
        current = current.parent

    return None


def analyze_deployment_location(skill_name: str, is_generic: bool = None) -> Dict:
    """Analyze and recommend deployment location"""
    global_dir = get_claude_skills_dir()
    project_dir = get_project_skills_dir()

    recommendations = {
        "skill_name": skill_name,
        "global_location": str(global_dir / skill_name),
        "project_location": str(project_dir / skill_name) if project_dir else None,
        "in_project": project_dir is not None,
        "recommendation": None,
        "reasons": []
    }

    # Determine recommendation
    if is_generic is True:
        # Explicitly marked as generic
        recommendations["recommendation"] = "global"
        recommendations["reasons"].append("Skill is marked as generic/reusable")
    elif is_generic is False:
        # Explicitly marked as project-specific
        recommendations["recommendation"] = "project"
        recommendations["reasons"].append("Skill is marked as project-specific")
    elif project_dir:
        # Auto-detect: if in a project, default to project
        recommendations["recommendation"] = "project"
        recommendations["reasons"].append("Currently in a project directory")
    else:
        # Not in a project, use global
        recommendations["recommendation"] = "global"
        recommendations["reasons"].append("Not in a project directory")

    # Check for existing installations
    if (global_dir / skill_name).exists():
        recommendations["existing_global"] = True
        recommendations["reasons"].append(f"⚠️  Skill already exists in global location")

    if project_dir and (project_dir / skill_name).exists():
        recommendations["existing_project"] = True
        recommendations["reasons"].append(f"⚠️  Skill already exists in project location")

    return recommendations


def check_dependencies() -> Dict:
    """Check Python version and common dependencies"""
    import sys

    checks = {
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}",
        "python_ok": sys.version_info >= (3, 8),
        "dependencies": {}
    }

    # Check for common packages
    common_packages = ["requests", "jinja2", "pyyaml"]

    for package in common_packages:
        try:
            __import__(package)
            checks["dependencies"][package] = "installed"
        except ImportError:
            checks["dependencies"][package] = "not installed"

    return checks


def extract_package(zip_path: Path, output_dir: Path) -> Path:
    """Extract skill package and return skill directory"""
    print(f"📦 Extracting package: {zip_path}")

    with zipfile.ZipFile(zip_path, 'r') as zipf:
        zipf.extractall(output_dir)

    # Find the skill directory (should be the only directory in the zip)
    skill_dirs = [d for d in output_dir.iterdir() if d.is_dir()]

    if len(skill_dirs) != 1:
        raise ValueError(f"Expected 1 skill directory, found {len(skill_dirs)}")

    return skill_dirs[0]


def deploy_skill(skill_path: Path, target_dir: Path, use_symlink: bool = False) -> bool:
    """Deploy skill to target directory"""
    skill_name = skill_path.name
    target_path = target_dir / skill_name

    # Check if target already exists
    if target_path.exists():
        response = input(f"⚠️  {target_path} already exists. Overwrite? (yes/no): ")
        if response.lower() != "yes":
            print("❌ Deployment cancelled")
            return False

        # Remove existing
        if target_path.is_symlink():
            target_path.unlink()
        else:
            shutil.rmtree(target_path)

    # Create target directory if needed
    target_dir.mkdir(parents=True, exist_ok=True)

    # Deploy
    if use_symlink:
        print(f"🔗 Creating symlink: {target_path} -> {skill_path}")
        target_path.symlink_to(skill_path.resolve())
    else:
        print(f"📁 Copying skill: {skill_path} -> {target_path}")
        shutil.copytree(skill_path, target_path)

    print(f"✅ Deployed to: {target_path}")
    return True


def analyze_package(zip_path: Path) -> Dict:
    """Analyze a skill package"""
    import tempfile

    with tempfile.TemporaryDirectory() as tmpdir:
        tmpdir_path = Path(tmpdir)

        # Extract to temp
        with zipfile.ZipFile(zip_path, 'r') as zipf:
            zipf.extractall(tmpdir_path)

        # Find metadata
        metadata_files = list(tmpdir_path.rglob(".skill-metadata.json"))

        if metadata_files:
            with open(metadata_files[0]) as f:
                metadata = json.load(f)
        else:
            metadata = {"error": "No metadata found"}

        # Find skill directory
        skill_dirs = [d for d in tmpdir_path.iterdir() if d.is_dir()]
        skill_name = skill_dirs[0].name if skill_dirs else "unknown"

        # Get deployment recommendation
        deployment_rec = analyze_deployment_location(skill_name)

        # Check dependencies
        dep_checks = check_dependencies()

        return {
            "metadata": metadata,
            "deployment": deployment_rec,
            "dependencies": dep_checks
        }


def main():
    parser = argparse.ArgumentParser(description="Deploy skills intelligently")
    parser.add_argument("skill_path", nargs='?', help="Path to skill directory or package")
    parser.add_argument("--scope", choices=["project", "global"],
                       help="Deployment scope")
    parser.add_argument("--symlink", action="store_true",
                       help="Create symlink instead of copying")
    parser.add_argument("--analyze", help="Analyze a package without deploying")

    args = parser.parse_args()

    # Analyze mode
    if args.analyze:
        zip_path = Path(args.analyze)
        if not zip_path.exists():
            print(f"❌ Package not found: {zip_path}")
            sys.exit(1)

        print("🔍 Analyzing package...")
        analysis = analyze_package(zip_path)

        print("\n" + "=" * 60)
        print("DEPLOYMENT ANALYSIS")
        print("=" * 60)
        print(json.dumps(analysis, indent=2))

        sys.exit(0)

    # Deploy mode
    if not args.skill_path:
        parser.print_help()
        sys.exit(1)

    skill_path = Path(args.skill_path)

    # Handle zip files
    if skill_path.suffix == '.zip':
        import tempfile
        tmpdir = tempfile.mkdtemp()
        skill_path = extract_package(skill_path, Path(tmpdir))

    if not skill_path.exists():
        print(f"❌ Skill path not found: {skill_path}")
        sys.exit(1)

    skill_name = skill_path.name

    # Get deployment recommendation
    recommendations = analyze_deployment_location(skill_name)

    print("\n" + "=" * 60)
    print(f"DEPLOYMENT PLANNING: {skill_name}")
    print("=" * 60)

    print(f"\n📍 Recommended scope: {recommendations['recommendation']}")
    for reason in recommendations['reasons']:
        print(f"  • {reason}")

    # Determine target
    if args.scope:
        scope = args.scope
    else:
        scope = recommendations['recommendation']

    if scope == "global":
        target_dir = get_claude_skills_dir()
    else:
        target_dir = get_project_skills_dir()
        if not target_dir:
            print("❌ Not in a project directory. Use --scope global")
            sys.exit(1)

    print(f"\n📁 Target: {target_dir}")

    # Check dependencies
    dep_checks = check_dependencies()
    print(f"\n🐍 Python {dep_checks['python_version']}: {'✅' if dep_checks['python_ok'] else '❌'}")

    # Deploy
    print(f"\n🚀 Deploying...")
    success = deploy_skill(skill_path, target_dir, args.symlink)

    if success:
        print(f"\n✅ Deployment complete!")
        print(f"\n📝 Test the skill by asking Claude:")
        print(f'   "Use the {skill_name} skill to [your task]"')

    sys.exit(0 if success else 1)


if __name__ == "__main__":
    main()

File: plugins/maencof/skills/craft-skill/scripts/deployment_helper.mjs
#!/usr/bin/env node
/**
 * Intelligent Deployment Helper - Assists with skill deployment decisions
 *
 * Usage:
 *     node deployment_helper.mjs <skill-path> [--scope project|global] [--symlink]
 *     node deployment_helper.mjs --analyze <package.zip>
 *
 * Features:
 * - Intelligent deployment location detection
 * - Project-specific vs global skill recommendations
 * - Symlink creation support
 * - Dependency checking
 */

import { parseArgs } from 'node:util';
import {
  existsSync, mkdirSync, rmSync, cpSync, symlinkSync,
  lstatSync, readdirSync, readFileSync,
} from 'node:fs';
import { execSync } from 'node:child_process';
import { createInterface } from 'node:readline/promises';
import { stdin, stdout } from 'node:process';
import path from 'node:path';
import { homedir, tmpdir } from 'node:os';
import { mkdtempSync } from 'node:fs';

function getClaudeSkillsDir() {
  return path.join(homedir(), '.claude', 'skills');
}

function getProjectSkillsDir() {
  let current = process.cwd();
  for (let i = 0; i < 5; i++) {
    const claudeDir = path.join(current, '.claude');
    if (existsSync(claudeDir)) {
      return path.join(claudeDir, 'skills');
    }
    const parent = path.dirname(current);
    if (parent === current) break;
    current = parent;
  }
  return null;
}

function analyzeDeploymentLocation(skillName, isGeneric = null) {
  const globalDir = getClaudeSkillsDir();
  const projectDir = getProjectSkillsDir();

  const recommendations = {
    skill_name: skillName,
    global_location: path.join(globalDir, skillName),
    project_location: projectDir ? path.join(projectDir, skillName) : null,
    in_project: projectDir !== null,
    recommendation: null,
    reasons: [],
  };

  if (isGeneric === true) {
    recommendations.recommendation = 'global';
    recommendations.reasons.push('Skill is marked as generic/reusable');
  } else if (isGeneric === false) {
    recommendations.recommendation = 'project';
    recommendations.reasons.push('Skill is marked as project-specific');
  } else if (projectDir) {
    recommendations.recommendation = 'project';
    recommendations.reasons.push('Currently in a project directory');
  } else {
    recommendations.recommendation = 'global';
    recommendations.reasons.push('Not in a project directory');
  }

  if (existsSync(path.join(globalDir, skillName))) {
    recommendations.existing_global = true;
    recommendations.reasons.push('⚠️  Skill already exists in global location');
  }

  if (projectDir && existsSync(path.join(projectDir, skillName))) {
    recommendations.existing_project = true;
    recommendations.reasons.push('⚠️  Skill already exists in project location');
  }

  return recommendations;
}

function checkDependencies() {
  const nodeVersion = process.version.replace('v', '');
  const [major] = nodeVersion.split('.').map(Number);

  return {
    node_version: nodeVersion,
    node_ok: major >= 18,
  };
}

function extractPackage(zipPath, outputDir) {
  console.log(`📦 Extracting package: ${zipPath}`);
  execSync(`unzip -o "${zipPath}" -d "${outputDir}"`, { stdio: 'pipe' });

  const skillDirs = readdirSync(outputDir, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .map((d) => path.join(outputDir, d.name));

  if (skillDirs.length !== 1) {
    throw new Error(`Expected 1 skill directory, found ${skillDirs.length}`);
  }

  return skillDirs[0];
}

async function deploySkill(skillPath, targetDir, useSymlink = false) {
  const skillName = path.basename(skillPath);
  const targetPath = path.join(targetDir, skillName);

  if (existsSync(targetPath)) {
    const rl = createInterface({ input: stdin, output: stdout });
    const response = await rl.question(`⚠️  ${targetPath} already exists. Overwrite? (yes/no): `);
    rl.close();

    if (response.toLowerCase() !== 'yes') {
      console.log('❌ Deployment cancelled');
      return false;
    }

    if (lstatSync(targetPath).isSymbolicLink()) {
      rmSync(targetPath);
    } else {
      rmSync(targetPath, { recursive: true, force: true });
    }
  }

  mkdirSync(targetDir, { recursive: true });

  if (useSymlink) {
    console.log(`🔗 Creating symlink: ${targetPath} -> ${skillPath}`);
    symlinkSync(path.resolve(skillPath), targetPath);
  } else {
    console.log(`📁 Copying skill: ${skillPath} -> ${targetPath}`);
    cpSync(skillPath, targetPath, { recursive: true });
  }

  console.log(`✅ Deployed to: ${targetPath}`);
  return true;
}

function analyzePackage(zipPath) {
  const tmpDir = mkdtempSync(path.join(tmpdir(), 'skill-'));

  try {
    execSync(`unzip -o "${zipPath}" -d "${tmpDir}"`, { stdio: 'pipe' });

    // Find metadata
    let metadata = { error: 'No metadata found' };
    const metadataPath = findFile(tmpDir, '.skill-metadata.json');
    if (metadataPath) {
      metadata = JSON.parse(readFileSync(metadataPath, 'utf-8'));
    }

    // Find skill directory
    const entries = readdirSync(tmpDir, { withFileTypes: true });
    const skillDirs = entries.filter((d) => d.isDirectory());
    const skillName = skillDirs.length > 0 ? skillDirs[0].name : 'unknown';

    const deploymentRec = analyzeDeploymentLocation(skillName);
    const depChecks = checkDependencies();

    return {
      metadata,
      deployment: deploymentRec,
      dependencies: depChecks,
    };
  } finally {
    rmSync(tmpDir, { recursive: true, force: true });
  }
}

function findFile(dir, filename) {
  const entries = readdirSync(dir, { withFileTypes: true });
  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isFile() && entry.name === filename) return fullPath;
    if (entry.isDirectory()) {
      const found = findFile(fullPath, filename);
      if (found) return found;
    }
  }
  return null;
}

const { values, positionals } = parseArgs({
  options: {
    scope: { type: 'string' },
    symlink: { type: 'boolean', default: false },
    analyze: { type: 'string' },
  },
  allowPositionals: true,
  strict: true,
});

// Analyze mode
if (values.analyze) {
  const zipPath = path.resolve(values.analyze);
  if (!existsSync(zipPath)) {
    console.error(`❌ Package not found: ${zipPath}`);
    process.exit(1);
  }

  console.log('🔍 Analyzing package...');
  const analysis = analyzePackage(zipPath);

  console.log('\n' + '='.repeat(60));
  console.log('DEPLOYMENT ANALYSIS');
  console.log('='.repeat(60));
  console.log(JSON.stringify(analysis, null, 2));
  process.exit(0);
}

// Deploy mode
if (positionals.length === 0) {
  console.error('Usage: node deployment_helper.mjs <skill-path> [--scope project|global] [--symlink]');
  console.error('       node deployment_helper.mjs --analyze <package.zip>');
  process.exit(1);
}

let skillPath = path.resolve(positionals[0]);

// Handle zip files
if (skillPath.endsWith('.zip')) {
  const tmpDir = mkdtempSync(path.join(tmpdir(), 'skill-'));
  skillPath = extractPackage(skillPath, tmpDir);
}

if (!existsSync(skillPath)) {
  console.error(`❌ Skill path not found: ${skillPath}`);
  process.exit(1);
}

const skillName = path.basename(skillPath);
const recommendations = analyzeDeploymentLocation(skillName);

console.log('\n' + '='.repeat(60));
console.log(`DEPLOYMENT PLANNING: ${skillName}`);
console.log('='.repeat(60));

console.log(`\n📍 Recommended scope: ${recommendations.recommendation}`);
for (const reason of recommendations.reasons) {
  console.log(`  • ${reason}`);
}

const scope = values.scope || recommendations.recommendation;

let targetDir;
if (scope === 'global') {
  targetDir = getClaudeSkillsDir();
} else {
  targetDir = getProjectSkillsDir();
  if (!targetDir) {
    console.error('❌ Not in a project directory. Use --scope global');
    process.exit(1);
  }
}

console.log(`\n📁 Target: ${targetDir}`);

const depChecks = checkDependencies();
console.log(`\n🟢 Node.js ${depChecks.node_version}: ${depChecks.node_ok ? '✅' : '❌'}`);

console.log('\n🚀 Deploying...');
const success = await deploySkill(skillPath, targetDir, values.symlink);

if (success) {
  console.log('\n✅ Deployment complete!');
  console.log(`\n📝 Test the skill by asking Claude:`);
  console.log(`   "Use the ${skillName} skill to [your task]"`);
}

process.exit(success ? 0 : 1);
